### PR TITLE
WIP: Call Selector Spread Priority in Default pod topology plugin

### LINKED ISSUE #86435: Move SelectorSpread priority logic to its Score plugin [Migration Phase 2]
/sig scheduling
/priority important-soon

SelectorSpread priority is now called from a plugin named DefaultPodTopologySpread. We need to move the logic to run in the plugin and remove SelectorSpread priority completely.

See #86399 as an example.

Part of #85822

This needs PostFilter Plugin to store the pod selector.

/assign @hprateek43

### PR DESCRIPTION
**Pull request by hprateek43**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #86435

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by hprateek43** [comment]
@ahg-g Can you review this?

File: pkg/scheduler/framework/plugins/defaultpodtopologyspread/default_pod_topology_spread.go [hunk 1]
```diff
@@ -86,3 +95,141 @@ func New(_ *runtime.Unknown, handle framework.FrameworkHandle) (framework.Plugin
...
+	if node == nil {
+		return framework.NodeScore{}, fmt.Errorf("node not found")
+	}
+
+	priorityMeta, ok := meta.(*priorityMetadata)
+	if ok {
+		selector = priorityMeta.podSelector
+	} else {
+		selector = getSelector(pod, s.serviceLister, s.controllerLister, s.replicaSetLister, s.statefulSetLister)
+	}
```
Comment (ahg-g) [inline pkg/scheduler/framework/plugins/defaultpodtopologyspread/default_pod_topology_spread.go]: this needs to be replaced with a PostFilter.

Comment (ahg-g) [reply pkg/scheduler/framework/plugins/defaultpodtopologyspread/default_pod_topology_spread.go]: see this plugin as an example: https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/framework/plugins/tainttoleration/taint_toleration.go#L84

File: pkg/scheduler/framework/plugins/defaultpodtopologyspread/default_pod_topology_spread.go [hunk 2]
```diff
@@ -86,3 +95,141 @@ func New(_ *runtime.Unknown, handle framework.FrameworkHandle) (framework.Plugin
...
+	count := countMatchingPods(pod.Namespace, selector, nodeInfo)
+	return framework.NodeScore{
+		Name:  node.Name,
+		Score: int64(count),
+	}, nil
+}
+
+// countMatchingPods counts pods based on namespace and matching all selectors
+func countMatchingPods(namespace string, selector labels.Selector, nodeInfo *schedulernodeinfo.NodeInfo) int {
+	if nodeInfo.Pods() == nil || len(nodeInfo.Pods()) == 0 || selector.Empty() {
```
Comment (ahg-g) [inline pkg/scheduler/framework/plugins/defaultpodtopologyspread/default_pod_topology_spread.go]: ```nodeInfo.Pods() == nil ||``` is redundant ```len(nodeInfo.Pods())``` returns zero when ```nodeInfo.Pods()``` is ```nil```

File: pkg/scheduler/framework/plugins/defaultpodtopologyspread/default_pod_topology_spread.go [hunk 3]
```diff
@@ -86,3 +95,141 @@ func New(_ *runtime.Unknown, handle framework.FrameworkHandle) (framework.Plugin
...
+				fScore = (fScore * (1.0 - zoneWeighting)) + (zoneWeighting * zoneScore)
+			}
+		}
+		result[i].Score = int64(fScore)
+		if klog.V(10) {
+			klog.Infof(
+				"%v -> %v: SelectorSpreadPriority, Score: (%d)", pod.Name, result[i].Name, int64(fScore),
+			)
+		}
+	}
```
Comment (ahg-g) [inline pkg/scheduler/framework/plugins/defaultpodtopologyspread/default_pod_topology_spread.go]: remove this.

**Review by ahg-g — COMMENTED** [review]
you also need to delete the code from algorithm/priorities/selector_spread[_test].go

File: pkg/scheduler/framework/plugins/defaultpodtopologyspread/default_pod_topology_spread.go [hunk 4]
```diff
@@ -86,3 +95,141 @@ func New(_ *runtime.Unknown, handle framework.FrameworkHandle) (framework.Plugin
 		calculateSpreadPriorityReduce: calculateSpreadPriorityReduce,
 	}, nil
 }
+
+//Migration Code
```
Comment (draveness) [inline pkg/scheduler/framework/plugins/defaultpodtopologyspread/default_pod_topology_spread.go]: nit: this should be removed

File: pkg/scheduler/framework/plugins/defaultpodtopologyspread/default_pod_topology_spread.go (hunk 1 above)
Comment (hprateek43) [reply pkg/scheduler/framework/plugins/defaultpodtopologyspread/default_pod_topology_spread.go]: Where should the PostFilter logic go? In this same file or in a different location?

**Review by hprateek43 — COMMENTED** [review]
@ahg-g Updated changes as per recommendations. Starting on the post-filter logic. I'm still unsure where the post-filter logic should go in the codebase.

File: pkg/scheduler/framework/plugins/defaultpodtopologyspread/default_pod_topology_spread.go (hunk 1 above)
Comment (ahg-g) [reply pkg/scheduler/framework/plugins/defaultpodtopologyspread/default_pod_topology_spread.go]: same file, just like the example I linked to.

File: pkg/scheduler/framework/plugins/defaultpodtopologyspread/default_pod_topology_spread.go [hunk 5]
```diff
@@ -86,3 +94,135 @@ func New(_ *runtime.Unknown, handle framework.FrameworkHandle) (framework.Plugin
...
+}
+
+// CalculateSpreadPriorityMap spreads pods across hosts, considering pods
+// belonging to the same service,RC,RS or StatefulSet.
+// When a pod is scheduled, it looks for services, RCs,RSs and StatefulSets that match the pod,
+// then finds existing pods that match those selectors.
+// It favors nodes that have fewer existing matching pods.
+// i.e. it pushes the scheduler towards a node where there's the smallest number of
+// pods which match the same service, RC,RSs or StatefulSets selectors as the pod being scheduled.
+func (s *SelectorSpread) CalculateSpreadPriorityMap(pod *v1.Pod, meta interface{}, nodeInfo *schedulernodeinfo.NodeInfo) (framework.NodeScore, error) {
```
Comment (ahg-g) [inline pkg/scheduler/framework/plugins/defaultpodtopologyspread/default_pod_topology_spread.go]: The logic in this function should go into the ```Score``` function.

File: pkg/scheduler/framework/plugins/defaultpodtopologyspread/default_pod_topology_spread.go [hunk 6]
```diff
@@ -86,3 +94,135 @@ func New(_ *runtime.Unknown, handle framework.FrameworkHandle) (framework.Plugin
...
+	controllerLister corelisters.ReplicationControllerLister,
+	replicaSetLister appslisters.ReplicaSetLister,
+	statefulSetLister appslisters.StatefulSetLister) (priorities.PriorityMapFunction, priorities.PriorityReduceFunction) {
+	selectorSpread := &SelectorSpread{
+		serviceLister:     serviceLister,
+		controllerLister:  controllerLister,
+		replicaSetLister:  replicaSetLister,
+		statefulSetLister: statefulSetLister,
+	}
+	return selectorSpread.CalculateSpreadPriorityMap, selectorSpread.CalculateSpreadPriorityReduce
```
Comment (ahg-g) [inline pkg/scheduler/framework/plugins/defaultpodtopologyspread/default_pod_topology_spread.go]: We don't need this type and function, we already have ```DefaultPodTopologySpread``` defined in line 36 and ```New``` at line 82

File: pkg/scheduler/framework/plugins/defaultpodtopologyspread/default_pod_topology_spread.go [hunk 7]
```diff
@@ -86,3 +94,135 @@ func New(_ *runtime.Unknown, handle framework.FrameworkHandle) (framework.Plugin
...
+		}
+	}
+	return count
+}
+
+// CalculateSpreadPriorityReduce calculates the source of each node
+// based on the number of existing matching pods on the node
+// where zone information is included on the nodes, it favors nodes
+// in zones with fewer existing matching pods.
+func (s *SelectorSpread) CalculateSpreadPriorityReduce(pod *v1.Pod, meta interface{}, sharedLister schedulerlisters.SharedLister, result framework.NodeScoreList) error {
```
Comment (ahg-g) [inline pkg/scheduler/framework/plugins/defaultpodtopologyspread/default_pod_topology_spread.go]: the logic in function should go into the ```NormalizeScore``` define at line 70

**Review by ahg-g — COMMENTED** [review]
Please take a look at the many previous similar PRs we merged that does the same thing for other priorities, here is an example one: https://github.com/kubernetes/kubernetes/pull/86585

other examples are under task 7 in the main issue: https://github.com/kubernetes/kubernetes/issues/85822

**Pull request closed without merging**